### ISSUE TITLE
Guide/Documentation for File Migrations

### ISSUE THREAD

**Issue by JeremyPlease** [parse-server-modules/parse-files-utils#10]
There's quite a few configuration options now and it's become a bit confusing 😞  Most people that use this tool will likely be migrating Parse.com hosted files to Parse Server and their own file adapter. 

Given ParsePlatform/parse-server#1582, there will not be any updates to api.parse.com. This leaves a few options for the file migration. Each one has its own set of advantages and disadvantages. I'll outline below for discussion. Then we can add some configuration samples for each one.

**Option 1**: 
"Supporting clients that access via api.parse.com is *not* important"
* File utils configuration:
  * `filesToTransfer`: 'parseOnly'
  * `renameFiles`: true
  * `renameInDatabase`: true
* Parse Server configuration:  keep using `fileKey` in settings
* After file migration:
  * api.parse.com clients:
    * can not see all previously uploaded files
    * can not see files uploaded by Parse Server clients
    * can see new files uploaded by api.parse.com clients
  * Parse Server clients:
    * can see all previously uploaded files
    * can see new files uploaded by api.parse.com clients
    * can see new files uploaded by Parse Server clients
* Additional steps required:
  * Run file migration again after all clients switch to Parse Server or before Jan 28

**Option 2**: 
"Supporting clients that access via api.parse.com is important"
* File utils configuration:
  * `filesToTransfer`: 'parseOnly'
  * `renameFiles`: false
  * `renameInDatabase`: false
* Parse Server configuration: 
  * Use version >= 2.2.16
  * remove `fileKey` from settings 
* After file migration:
  * api.parse.com clients:
    * can see all previously uploaded files
    * can not see files uploaded by Parse Server clients
  * Parse Server clients:
    * can see all migrated files
    * can not see new files uploaded by api.parse.com clients
* Additional steps required:
  * Run file migration again after all clients switch to Parse Server or before Jan 28

---

Well, I guess there are only two logical options. Anyone have any thoughts, questions, or feedback? What do you think the best way to document this is (wiki, readme, diagrams)?

**Comment by acinader** [comment]
typo in option 1:
> Parse Server clients:
> * can see all previously uploaded files
> * can see new files uploaded by api.parse.com clients
> * can see new files uploaded by api.parse.com clients

one of the last two lines should be ```can see new files uploaded by Parse Server clients```

I think it is clear and makes clear what the issues are.  is helpful to me.  thanks.  as for how to document it, i think what you have above is a good start as is.

**Comment by natanrolnik** [comment]
In the option 1, instead of:
> Parse Server configuration: no change
I would write explicitly:
> Parse Server configuration: keep using `fileKey`

**Comment by JeremyPlease** [comment]
Option 1 updated--thanks! 

I'll let this sit the rest of the day. If I come up with a nifty way to document or diagram then I'll update here. If not, then I'll put this in a separate section in the readme.

**Comment by acinader** [comment]
ok, here's another question: in the case of option 2, when we re-run the script, what is the expected behavior wrt files that were transferred in past runs?

i'm assuming (off hand) that it is adapter dependent.  and what should the behavior be.  My take would be that if a file with the same name already exists on the destination, then do nothing and move on.

**Comment by JeremyPlease** [comment]
@acinader The migration tool is just going to pass the filename and data to the file adapter. I'm pretty sure they all just replace existing files with new ones that have the same name.

Hypothetically, we check to see if a file exists in destination before uploading but that would be *another* configuration option 😑  I say it's best to leave that up to the file adapters.

**Comment by acinader** [comment]
thanks.  works for me.

**Comment by acinader** [comment]
I tested rename option, worked as expected.  Also as a bonus, my internet failed so I had to restart the migration and that worked just fine too.  Onto the no rename option.

**Comment by Brownmonster** [comment]
Great set of scripts, I have > 30000 entries in my database,  but the query only returns the first 1000 results and gets the files for them, I've had a quick go at changing this behaviour locally but to no success yet (I'll try later on a bit more). Is this a known issue for you?

**Comment by acinader** [comment]
@Brownmonster when you say that you have greater than 30,000 entries, do mean you have that many classes (i.e. Users, Installations, etc.) or do you mean you have a class with more than 30k records?

If its the former, then we've got a problem.  If it's the later, then it should be no problem.

**Comment by Brownmonster** [comment]
I've added some logging locally and it actually all appears to work going through the whole database, the problem for me was that in the initial query it only fetches rows where EVERY file exists in that row, which is not the case in my database.

I simply removed the check and it works fine :)

**Comment by acinader** [comment]
do you mean that you made a change here: https://github.com/parse-server-modules/parse-files-utils/blob/master/lib/transfer.js#L167-L181

Can push your change to a fork so I can check it out?  I'd like to know what the problem is, so if it is general we can fix this script.

**Comment by JeremyPlease** [comment]
@Brownmonster 
Thanks for testing the use case of multiple files in one object where all files are not set. I'm pretty sure you're referring to [this code](https://github.com/parse-server-modules/parse-files-utils/blob/master/lib/index.js#L93-L95) that checks to make sure each field `exists`.

Could you please open a separate issue and/or pull request about this?

**Comment by Brownmonster** [comment]
@JeremyPlease Yes thats the code, however it needs a few more changes
@acinader i havent touched that file no.

FYI, this is what i changed, it basically gets every file that is in use by the database and doesnt rely on all files existing in a row, it has abit more output too so you know what it is doing. Takes about an hour to run on my database of 30000 files, its not the most efficient bit of js, but who cares as it works!

I'm actually really new to this git stuff, i'm pretty much a windows/ios coder who uses other means so I'm happy to push this your way if you show me how :)

[index.txt](https://github.com/parse-server-modules/parse-files-utils/files/388771/index.txt)

**Comment by Brownmonster** [comment]
I've also had a few other issues where if the server drops out, the whole thing will terminate (in the case of a request (error), where error != null. I've stopped this from happening too as it was stopping half way through 250000 files!

I've also added a check to see if the file already exists, this means I can keep running it every day alongside a parse migrate to keep the database up to date until the final switch over.

Basically changed the bit at the bottom here in this txt file (not sure how to upload js as it wont let me and i'm new to all this!)

[transfer.txt](https://github.com/parse-server-modules/parse-files-utils/files/390329/transfer.txt)

**Comment by acinader** [comment]
Hey Stephen, nice.

Can I help you make a pull request?  If you're gonna keep using parse, you'll want to stay current and it'll be a lot easier if you can push you're changes for inclusion.  It's also easy and I'm glad to answer any questions along the way.  Here's the docs from github https://help.github.com/articles/fork-a-repo/

And a quick overview:

1. create your fork of this repo by going to the homepage of the project on git: https://github.com/parse-server-modules/parse-files-utils and clicking the fork button on the upper right:
<img width="394" alt="screen shot 2016-07-29 at 6 48 07 am" src="https://cloud.githubusercontent.com/assets/700572/17246220/a7ded324-5558-11e6-8604-960b82c3d103.png">

2. copy the url to clone your fork:
<img width="458" alt="screen shot 2016-07-29 at 6 50 58 am" src="https://cloud.githubusercontent.com/assets/700572/17246266/ec14b928-5558-11e6-8c62-9bb38c7b957e.png">

3. add your fork to your local repo.  Open a terminal and switch into the directory that has the files and type:
```git remote add upstream git@github.com:Brownmonster/parse-files-utils.git```

4. create a branch:
```git checkout -b dont-reupload-existsing-files```

5. commit your changes:
```git commit -a```

6. push your changes to your branch:
```git push upstream dont-reupload-existing-files```

7. go back to the homepage of the project on github: https://github.com/parse-server-modules/parse-files-utils

and it should prompt you to create a pull request.
